crucible-llvm/src/Lang/Crucible/LLVM/Intrinsics/Libc.hs

```diff
@@ -782,6 +824,84 @@ callBSwap ::
+data CheckAbsIntMin
+  = LibcAbsIntMinUB
+    -- ^ For the @abs@, @labs@, and @llabs@ functions, always check if the
+    --   argument is equal to @INT_MIN@. If so, report it as undefined
+    --   behavior per the C standard.
+  | LLVMAbsIntMinPoison Bool
+    -- ^ For the @llvm.abs.*@ family of LLVM intrinsics, check if the argument
+    --   is equal to @INT_MIN@ only when the 'Bool' argument is 'True'. If it
+    --   is 'True' and the argument is equal to @INT_MIN@, return poison.
+
+-- | The workhorse for the @abs@, @labs@, and @llabs@ functions, as well as the
+-- @llvm.abs.*@ family of overloaded intrinsics.
+callAbs ::
+  forall w p sym ext r args ret.
+  (1 <= w, IsSymInterface sym, HasLLVMAnn sym) =>
+  sym ->
+  CheckAbsIntMin ->
+  NatRepr w ->
+  RegEntry sym (BVType w) ->
+  OverrideSim p sym ext r args ret (RegValue sym (BVType w))
+callAbs sym checkIntMin widthRepr (regValue -> src) = liftIO $ do
+  bvIntMin      <- bvLit sym widthRepr (BV.minSigned widthRepr)
+  isNotIntMin   <- notPred sym =<< bvEq sym src bvIntMin
+  isSrcNegative <- bvIsNeg sym src
+  srcNegated    <- bvNeg sym src
+  srcAbs        <- bvIte sym isSrcNegative srcNegated src
+  if shouldCheckIntMin
+     then do isNotIntMinUB <- annotateUB sym ub isNotIntMin
+             let err = AssertFailureSimError "Undefined behavior encountered" $
+                       show $ UB.explain ub
+             assert sym isNotIntMinUB err
+             pure srcAbs
+     else -- Per https://releases.llvm.org/12.0.0/docs/LangRef.html#id452,
+          -- return INT_MIN if the argument is INT_MIN and we are not returning
+          -- poison.
+          bvIte sym isNotIntMin srcAbs src
+  where
+    shouldCheckIntMin :: Bool
+    shouldCheckIntMin =
+      case checkIntMin of
+        LibcAbsIntMinUB                 -> True
+        LLVMAbsIntMinPoison shouldCheck -> shouldCheck
+
+    ub :: UB.UndefinedBehavior (RegValue' sym)
+    ub = case checkIntMin of
+           LibcAbsIntMinUB ->
+             UB.AbsIntMin $ RV src
+           LLVMAbsIntMinPoison{} ->
+             UB.PoisonValueCreated $ Poison.LLVMAbsIntMin $ RV src
+
+callLibcAbs ::
+  (1 <= w, IsSymInterface sym, HasLLVMAnn sym) =>
+  sym ->
+  NatRepr w ->
+  RegEntry sym (BVType w) ->
+  OverrideSim p sym ext r args ret (RegValue sym (BVType w))
+callLibcAbs sym = callAbs sym LibcAbsIntMinUB
+
+callLLVMAbs ::
+  (1 <= w, IsSymInterface sym, HasLLVMAnn sym) =>
+  sym ->
+  NatRepr w ->
+  RegEntry sym (BVType w) ->
+  RegEntry sym (BVType 1) ->
+  OverrideSim p sym ext r args ret (RegValue sym (BVType w))
+callLLVMAbs sym widthRepr src (regValue -> isIntMinPoison) = do
+  shouldCheckIntMin <- liftIO $
+    -- Per https://releases.llvm.org/12.0.0/docs/LangRef.html#id451, the second
+    -- argument must be a constant.
+    case asBV isIntMinPoison of
+      Just bv -> pure (bv /= BV.zero (knownNat @1))
+      Nothing -> addFailedAssertion sym "Call to llvm.abs.* with non-constant second argument"
```

Comment:
This might be a case where we would want to throw [`MalformedLLVMModule`](https://github.com/GaloisInc/crucible/blob/master/crucible-llvm/src/Lang/Crucible/LLVM/MalformedLLVMModule.hs) instead of adding an assertion - It's not really a safety condition violation, but more of an issue that the module isn't semantically meaningful if this isn't a constant.

It may also be worth considering adding a check to LLVM module translation that the argument is syntactically constant, in order to fail earlier/more eagerly.

Reply:
Good suggestion about `MalformedLLVMModule`. I'll change this code to throw that instead.

I don't have a strong opinion about whether we check for the second argument being constant here or during module translation. Do you know if there are examples of other parts of the translation code where we check for constants in function arguments?

Reply:
I've pushed a change such that we now error out with `malformedLLVMModule "Call to llvm.abs.* with non-constant second argument" [printSymExpr isIntMinPoison]` instead.